### PR TITLE
Add more documentation do Syntax nodes

### PR DESCRIPTION
**Pull request by kimdv**
Part of https://github.com/apple/swift-syntax/issues/1528

I've started to add more documentation to the syntax nodes. 
Some of them are pretty trivial and not sure if that is the path we want. 

Like the effect specifiers, we could generate them if we wan't them. 
I think they give some value specially places where we use `asyncSpecifier` etc. Here are two valid options. 
We could generate 
```
The `async` or `reasync` keyword.
```

Then they could be able to click on them and navigate to it. 

When generating this, we could make rule, that if we define the documentation then it overrules the generated ones

### CODE REVIEW

File: CodeGeneration/Sources/SyntaxSupport/CommonNodes.swift [hunk 1]
```diff
@@ -114,11 +118,13 @@ public let COMMON_NODES: [Node] = [
       Child(
         name: "asyncSpecifier",
         kind: .token(choices: [.keyword(.async), .keyword(.reasync)]),
+        documentation: "The `async` or `reasync` keyword.",
         isOptional: true
       ),
       Child(
         name: "throwsSpecifier",
         kind: .token(choices: [.keyword(.throws), .keyword(.rethrows)]),
+        documentation: "The `throws` or `rethrows` keyword",
```
Comment (SimplyDanny) [inline CodeGeneration/Sources/SyntaxSupport/CommonNodes.swift:127]:
```suggestion
        documentation: "The `throws` or `rethrows` keyword.",
```

File: CodeGeneration/Sources/SyntaxSupport/DeclNodes.swift [hunk 2]
```diff
@@ -53,7 +53,8 @@ public let DECL_NODES: [Node] = [
     children: [
       Child(
         name: "leftBrace",
-        kind: .token(choices: [.token(.leftBrace)])
+        kind: .token(choices: [.token(.leftBrace)]),
+        documentation: "The brace that is the opening of the accessor block"
```
Comment (SimplyDanny) [inline CodeGeneration/Sources/SyntaxSupport/DeclNodes.swift:57]:
```suggestion
        documentation: "The brace introducing the accessor block."
```

File: CodeGeneration/Sources/SyntaxSupport/DeclNodes.swift [hunk 3]
```diff
@@ -70,7 +71,8 @@ public let DECL_NODES: [Node] = [
       ),
       Child(
         name: "rightBrace",
-        kind: .token(choices: [.token(.rightBrace)])
+        kind: .token(choices: [.token(.rightBrace)]),
+        documentation: "The brace that is the closing of the accessor block"
```
Comment (SimplyDanny) [inline CodeGeneration/Sources/SyntaxSupport/DeclNodes.swift:75]:
```suggestion
        documentation: "The brace closing the accessor block."
```

File: CodeGeneration/Sources/SyntaxSupport/DeclNodes.swift [hunk 4]
```diff
@@ -200,7 +203,8 @@ public let DECL_NODES: [Node] = [
       Child(
         name: "name",
         deprecatedName: "identifier",
-        kind: .token(choices: [.token(.identifier)])
+        kind: .token(choices: [.token(.identifier)]),
+        documentation: "The identifier of the actor"
```
Comment (SimplyDanny) [inline CodeGeneration/Sources/SyntaxSupport/DeclNodes.swift:207]:
```suggestion
        documentation: "The name of the actor."
```

File: CodeGeneration/Sources/SyntaxSupport/DeclNodes.swift [hunk 5]
```diff
@@ -857,7 +861,8 @@ public let DECL_NODES: [Node] = [
       Child(
         name: "modifiers",
         kind: .collection(kind: .declModifierList, collectionElementName: "Modifier", defaultsToEmpty: true),
-        nameForDiagnostics: "modifiers"
+        nameForDiagnostics: "modifiers",
+        documentation: "Modifiers modifiers applied to the extension declaration."
```
Comment (SimplyDanny) [inline CodeGeneration/Sources/SyntaxSupport/DeclNodes.swift:865]:
```suggestion
        documentation: "Modifiers attached to the extension declaration."
```

File: CodeGeneration/Sources/SyntaxSupport/DeclNodes.swift [hunk 6]
```diff
@@ -191,7 +193,8 @@ public let DECL_NODES: [Node] = [
       Child(
         name: "modifiers",
         kind: .collection(kind: .declModifierList, collectionElementName: "Modifier", defaultsToEmpty: true),
-        nameForDiagnostics: "modifiers"
+        nameForDiagnostics: "modifiers",
+        documentation: "Modifiers attached to the actor."
```
Comment (SimplyDanny) [inline CodeGeneration/Sources/SyntaxSupport/DeclNodes.swift:197]:
```suggestion
        documentation: "Modifiers attached to the actor declaration."
```

File: CodeGeneration/Sources/SyntaxSupport/DeclNodes.swift [hunk 7]
```diff
@@ -906,7 +911,8 @@ public let DECL_NODES: [Node] = [
       Child(
         name: "modifiers",
         kind: .collection(kind: .declModifierList, collectionElementName: "Modifier", defaultsToEmpty: true),
-        nameForDiagnostics: "modifiers"
+        nameForDiagnostics: "modifiers",
+        documentation: "Modifiers modifiers applied to the function declaration."
```
Comment (SimplyDanny) [inline CodeGeneration/Sources/SyntaxSupport/DeclNodes.swift:915]:
```suggestion
        documentation: "Modifiers attached to the function declaration."
```

File: CodeGeneration/Sources/SyntaxSupport/DeclNodes.swift [hunk 8]
```diff
@@ -2324,7 +2330,8 @@ public let DECL_NODES: [Node] = [
       Child(
         name: "modifiers",
         kind: .collection(kind: .declModifierList, collectionElementName: "Modifier", defaultsToEmpty: true),
-        nameForDiagnostics: "modifiers"
+        nameForDiagnostics: "modifiers",
+        documentation: "Modifiers modifiers applied to the variable declaration."
```
Comment (SimplyDanny) [inline CodeGeneration/Sources/SyntaxSupport/DeclNodes.swift:2334]:
```suggestion
        documentation: "Modifiers attached to the variable declaration."
```

File: CodeGeneration/Sources/SyntaxSupport/ExprNodes.swift [hunk 9]
```diff
@@ -1683,37 +1683,50 @@ public let EXPR_NODES: [Node] = [
...
+    documentation: """
+      A `switch` statement expression.
+
+      This node represents both a 'switch' expression, as well as a 'switch'
+      statement when wrapped in an ``ExpressionStmtSyntax`` node.
+
+      ```swift
+      switch <#expression#> {
+      }
+      ```
```
Comment (SimplyDanny) [inline CodeGeneration/Sources/SyntaxSupport/ExprNodes.swift:1699]: What is this snippet supposed to mean?

File: CodeGeneration/Sources/SyntaxSupport/ExprNodes.swift [hunk 10]
```diff
@@ -1683,37 +1683,50 @@ public let EXPR_NODES: [Node] = [
...
+      """,
     traits: [
       "Braced"
     ],
     children: [
       Child(
         name: "switchKeyword",
-        kind: .token(choices: [.keyword(.switch)])
+        kind: .token(choices: [.keyword(.switch)]),
+        documentation: "The switch keyword."
```
Comment (SimplyDanny) [inline CodeGeneration/Sources/SyntaxSupport/ExprNodes.swift:1708]:
```suggestion
        documentation: "The `switch` keyword."
```

File: CodeGeneration/Sources/SyntaxSupport/ExprNodes.swift [hunk 11]
```diff
@@ -1683,37 +1683,50 @@ public let EXPR_NODES: [Node] = [
...
         deprecatedName: "expression",
-        kind: .node(kind: .expr)
+        kind: .node(kind: .expr),
+        documentation: "The expression to switch over."
       ),
       Child(
         name: "leftBrace",
-        kind: .token(choices: [.token(.leftBrace)])
+        kind: .token(choices: [.token(.leftBrace)]),
+        documentation: "The left brace of the switch body."
```
Comment (SimplyDanny) [inline CodeGeneration/Sources/SyntaxSupport/ExprNodes.swift:1719]:
```suggestion
        documentation: "The brace introducing the switch body."
```

File: CodeGeneration/Sources/SyntaxSupport/ExprNodes.swift [hunk 12]
```diff
@@ -1683,37 +1683,50 @@ public let EXPR_NODES: [Node] = [
...
       ),
       Child(
         name: "cases",
-        kind: .collection(kind: .switchCaseList, collectionElementName: "Case")
+        kind: .collection(kind: .switchCaseList, collectionElementName: "Case"),
+        documentation: "The switch's body that contains the possible cases."
```
Comment (SimplyDanny) [inline CodeGeneration/Sources/SyntaxSupport/ExprNodes.swift:1724]:
```suggestion
        documentation: "The switch's body that contains all possible cases."
```

File: CodeGeneration/Sources/SyntaxSupport/ExprNodes.swift [hunk 13]
```diff
@@ -1683,37 +1683,50 @@ public let EXPR_NODES: [Node] = [
...
       ),
       Child(
         name: "rightBrace",
-        kind: .token(choices: [.token(.rightBrace)])
+        kind: .token(choices: [.token(.rightBrace)]),
+        documentation: "The right brace of the switch body."
```
Comment (SimplyDanny) [inline CodeGeneration/Sources/SyntaxSupport/ExprNodes.swift:1729]:
```suggestion
        documentation: "The brace closing the switch body."
```

**Review by SimplyDanny — COMMENTED** [review]
Just a few suggestions from my side ...

File: CodeGeneration/Sources/SyntaxSupport/ExprNodes.swift (hunk 9 above)
Comment (kimdv) [reply CodeGeneration/Sources/SyntaxSupport/ExprNodes.swift:1699]: The idea is to have a code-snip so the user can see an example if what the output is? 
Maybe it's just overhead. Not sure.

The idea comes from 
https://github.com/apple/swift-syntax/blob/9e798fa61ad386d9b59efe3d558fe1a734aa16c9/CodeGeneration/Sources/SyntaxSupport/DeclNodes.swift#L1232-L1239

Comment (SimplyDanny) [reply CodeGeneration/Sources/SyntaxSupport/ExprNodes.swift:1699]: I'm was only a bit confused by the `<#expression#>` which immediately caught my attention. Thought it was the thing to look at principally. A more concrete example like in the initializer example would be better.

File: CodeGeneration/Sources/SyntaxSupport/DeclNodes.swift [hunk 14]
```diff
@@ -191,16 +193,19 @@ public let DECL_NODES: [Node] = [
       Child(
         name: "modifiers",
         kind: .collection(kind: .declModifierList, collectionElementName: "Modifier", defaultsToEmpty: true),
-        nameForDiagnostics: "modifiers"
+        nameForDiagnostics: "modifiers",
+        documentation: "Modifiers that are attached to the actor declaration."
```
Comment (ahoppen) [inline CodeGeneration/Sources/SyntaxSupport/DeclNodes.swift:197]:
```suggestion
        documentation: "Modifiers like `public` that are attached to the actor declaration."
```

Same for the other documentation of modifiers.

File: CodeGeneration/Sources/SyntaxSupport/ExprNodes.swift [hunk 15]
```diff
@@ -1683,37 +1683,50 @@ public let EXPR_NODES: [Node] = [
...
   //   switch-case-list '}' ';'?
-  //
-  // This node represents both a 'switch' expression, as well as a 'switch'
-  // statement when wrapped in an ExpressionStmt node.
   Node(
     kind: .switchExpr,
     base: .expr,
     nameForDiagnostics: "'switch' statement",
+    documentation: """
+      A `switch` statement expression.
```
Comment (ahoppen) [inline CodeGeneration/Sources/SyntaxSupport/ExprNodes.swift:1691]:
```suggestion
      A `switch` expression.
```

File: CodeGeneration/Sources/SyntaxSupport/ExprNodes.swift [hunk 16]
```diff
@@ -1683,37 +1683,50 @@ public let EXPR_NODES: [Node] = [
...
+
+      This node represents both a 'switch' expression, as well as a 'switch'
+      statement when wrapped in an ``ExpressionStmtSyntax`` node.
```
Comment (ahoppen) [inline CodeGeneration/Sources/SyntaxSupport/ExprNodes.swift:1694]: I would just talk about switch expressions here. Technically, there are no switch statements in SwiftSyntax.

File: CodeGeneration/Sources/SyntaxSupport/ExprNodes.swift (hunk 9 above)
Comment (ahoppen) [reply CodeGeneration/Sources/SyntaxSupport/ExprNodes.swift:1699]: I agree that a concrete example would be better here instead of the placeholder.

**Review by ahoppen — COMMENTED** [review]
Mostly looks good to me. Thanks @kimdv. I wonder if cases like

```
The `actor` keyword
```

can be handled as part of https://github.com/apple/swift-syntax/issues/1987

CC @natikgadzhi

**Comment by natikgadzhi** [comment]
@ahoppen @kimdv yep! I haven't jumped into #1987 fully yet, but it does feel like when a child has just one / few possible _kinds_ or _specific tokens_ — we could generate documentation on those automatically based on those kinds, without hard-coding the specific doc string for each child. Thank you for nudging me here! <3

**Comment by kimdv** [comment]
> Mostly looks good to me. Thanks @kimdv. I wonder if cases like
> 
> ```
> The `actor` keyword
> ```
> 
> can be handled as part of #1987
> 
> CC @natikgadzhi

Will remove those cases again. 
Then it can be handled in #1987.

File: CodeGeneration/Sources/SyntaxSupport/ExprNodes.swift [hunk 17]
```diff
@@ -1683,37 +1683,51 @@ public let EXPR_NODES: [Node] = [
...
-  // This node represents both a 'switch' expression, as well as a 'switch'
-  // statement when wrapped in an ExpressionStmt node.
   Node(
     kind: .switchExpr,
     base: .expr,
     nameForDiagnostics: "'switch' statement",
+    documentation: """
+      A `switch` expression.
+
+      This node represents both a 'switch' expression.
```
Comment (ahoppen) [inline CodeGeneration/Sources/SyntaxSupport/ExprNodes.swift:1693]: I would just remove this sentence.

File: CodeGeneration/Sources/SyntaxSupport/ExprNodes.swift [hunk 18]
```diff
@@ -1683,37 +1683,51 @@ public let EXPR_NODES: [Node] = [
...
+
+      ```swift
+      switch self.foo {
+      }
+      ```
```
Comment (ahoppen) [inline CodeGeneration/Sources/SyntaxSupport/ExprNodes.swift:1698]: In https://github.com/apple/swift-syntax/pull/2179, I’m introducing an `### Examples` section to a few nodes, which would be a good fit here as well, I think.

**Comment by kimdv** [comment]
@ahoppen added more documentation. 
Should we review and merge this or should I add more?

Can also add more in a separate PR if you want?

File: CodeGeneration/Package.swift
Comment (ahoppen) [inline CodeGeneration/Package.swift:1]: Looks like this change is unrelated.

**Review by ahoppen — APPROVED** [review]
Yes, let’s merge this. Just looks like the change to CodeGeneration/Package.swift is unrelated.

**Comment by kimdv** [comment]
@swift-ci please test

**Comment by kimdv** [comment]
@swift-ci please test windows
